connect/runtime/src/main/java/org/apache/kafka/connect/storage/OffsetStorageReaderImpl.java

```diff
@@ -108,4 +140,14 @@ public OffsetStorageReaderImpl(OffsetBackingStore backingStore, String namespace
 
         return result;
     }
+
+    public void close() {
+        synchronized (offsetReadFutures) {
+            closed.set(true);
+            for (Future<Map<ByteBuffer, ByteBuffer>> offsetReadFuture : offsetReadFutures) {
+                offsetReadFuture.cancel(true);
```

Comment:
It is now possible with your changes for `ConvertingFutureCallback` to throw an exception during `cancel(boolean)`. If that happens, then this reader instance will not be properly/completely closed, so that needs to be handled.

Reply:
Ack, will address.